File: web/lib/db/prompt-versions.ts
import { createClient } from "@/lib/supabase/server";
import type { Database } from "@/lib/db/types";

type PromptVersionRow =
  Database["public"]["Tables"]["prompt_versions"]["Row"];

export async function createPromptVersion(
  workspaceId: string,
  systemPrompt: string,
  createdBy: string
): Promise<PromptVersionRow> {
  const supabase = await createClient();

  const { data: latest } = await supabase
    .from("prompt_versions")
    .select("version_number")
    .eq("workspace_id", workspaceId)
    .order("version_number", { ascending: false })
    .limit(1)
    .single();

  const nextVersion = latest ? latest.version_number + 1 : 1;

  const { data, error } = await supabase
    .from("prompt_versions")
    .insert({
      workspace_id: workspaceId,
      version_number: nextVersion,
      system_prompt: systemPrompt,
      created_by: createdBy,
    })
    .select()
    .single();

  if (error) throw error;
  return data;
}

export async function getLatestPromptVersion(
  workspaceId: string
): Promise<PromptVersionRow | null> {
  const supabase = await createClient();
  const { data, error } = await supabase
    .from("prompt_versions")
    .select("*")
    .eq("workspace_id", workspaceId)
    .order("version_number", { ascending: false })
    .limit(1)
    .single();
  if (error) return null;
  return data;
}

File: worker/db/prompt_versions.py
from typing import Any

from supabase import AsyncClient


async def create_prompt_version(
    client: AsyncClient,
    workspace_id: str,
    system_prompt: str,
    created_by: str,
    source: str = "manual",
) -> dict[str, Any]:
    latest = (
        await client.table("prompt_versions")
        .select("version_number")
        .eq("workspace_id", workspace_id)
        .order("version_number", desc=True)
        .limit(1)
        .maybe_single()
        .execute()
    )
    next_version = (latest.data["version_number"] + 1) if latest.data else 1

    res = (
        await client.table("prompt_versions")
        .insert(
            {
                "workspace_id": workspace_id,
                "version_number": next_version,
                "system_prompt": system_prompt,
                "created_by": created_by,
                "source": source,
            }
        )
        .execute()
    )
    return res.data[0]
